cmd/clusterctl/client/cluster/topology.go

```diff
@@ -214,13 +218,23 @@ func (t *topologyClient) DryRun(in *DryRunInput) (*DryRunOutput, error) {
 	return res, nil
 }
 
-// validateInput checks that the dryrun input does not violate any of the below expectations:
+// validateInput checks that the topology plan input does not violate any of the below expectations:
 // - no more than 1 cluster in the input.
 // - no more than 1 clusterclass in the input.
-func (t *topologyClient) validateInput(in *DryRunInput) error {
+func (t *topologyClient) validateInput(in *TopologyPlanInput) error {
+	// Check all the objects in the input belong to the same namespace.
+	// Note: It is okay if all the objects in the input do not have any namespace.
+	// In such case, the list of unique namespaces will be [""].
 	if len(uniqueNamespaces(in.Objs)) != 1 {
 		return fmt.Errorf("all the objects in the input should belong to the same namespace")
 	}
+
+	ns := uniqueNamespaces(in.Objs)[0]
+	// If the objects have a non empty namespace make sure that is matches the TargetNamespace.
+	if ns != "" && in.TargetNamespace != "" && ns != in.TargetNamespace {
+		return fmt.Errorf("the namespace from the provided object(s) %q does not match the namespace %q", ns, in.TargetNamespace)
+	}
+
```

Comment:
What if we set in.TargetNamespace to ns if not aready set, so we can avoid calling unique namespace in affectedClusterClasses and affectedClusters?
Also, this have the nice effect to keep the namespace logic in one single point